includes/class-wcsr-resource.php

```diff
@@ -187,17 +188,26 @@ public function get_days_active( $from_timestamp, $to_timestamp = null ) {
 		$activation_times   = self::get_timestamps_between( $this->get_activation_timestamps(), $from_timestamp, $to_timestamp );
 		$deactivation_times = self::get_timestamps_between( $this->get_deactivation_timestamps(), $from_timestamp, $to_timestamp );
 
-		// Now find the number of days between the timestamps where the resource was active (the resource must be inactive for more than 24 hours to be considered inactive for a given day)
-		$days_active = floor( ( $to_timestamp - $from_timestamp ) / DAY_IN_SECONDS );
-
-		// First remove any gap between the $from_timestamp and the resources creation date
-		if ( $this->get_date_created()->getTimestamp() > $from_timestamp ) {
-			$days_active -= floor( ( $this->get_date_created()->getTimestamp() - $from_timestamp ) / DAY_IN_SECONDS );
+		// if the first activation date is after the first deactivation date, make sure we append the start timestamps to act as the first "activated" date for the resource
+		if ( ! isset( $activation_times[0] ) || ( isset( $deactivation_times[0] ) && $activation_times[0] > $deactivation_times[0] ) ) {
+			$start_timestamp = ( $this->get_date_created()->getTimestamp() > $from_timestamp ) ? $this->get_date_created()->getTimestamp() : $from_timestamp;
+			array_unshift( $activation_times, $start_timestamp );
 		}
 
 		foreach ( $activation_times as $i => $activation_time ) {
-			if ( isset( $deactivation_times[ $i ] ) ) {
-				$days_active -= floor( ( $activation_time - $deactivation_times[ $i ] ) / DAY_IN_SECONDS );
+			// If there is corresponding deactivation timestamp, the resouce has deactivated before the end of the period so that's the time we want, otherwise, use the end of the period as the resource was still active at end of the period
+			$deactivation_time = isset( $deactivation_times[ $i ] ) ? $deactivation_times[ $i ] : $to_timestamp;
+
+			// skip over any days that are activated/deactivated on the same day and have already been accounted for
+			if ( $i !== 0 && gmdate( 'Y-m-d', $deactivation_times[ $i - 1 ] ) == gmdate( 'Y-m-d', $deactivation_time ) ) {
+				continue;
+			}
+
+			$days_active += intval( ceil( ( $deactivation_time - $activation_time ) / DAY_IN_SECONDS ) );
+
+			// if the activation date is the same as the previous activation date, minus one off one day active from the result since that day was already accounted for previously
+			if ( $i !== 0 && gmdate( 'Y-m-d', $activation_times[ $i - 1 ] ) == gmdate( 'Y-m-d', $activation_times[ $i ] ) ) {
```

Comment:
If I remove this check - all the tests pass except for 15 https://cl.ly/0v3c1e3e112a

Looking at 15 it is expecting 1 day - but I think it should probably be 2 as it is activated across 2 days from what I can tell based on the dates - in which case it will pass as well.

It's quite possible I am reading it wrong - I'm thinking if you look at 15 (ignoring the doc block)

Activated from 2017-09-14 20:00:03 to 2017-09-14 21:00:03 (1 hour) then reactivated an hour later from 2017-09-14 22:00:03 to 2017-09-15 01:15:11 (3 hours) - spanning into the next day - so 2 days? (cycle finishes on 2017-10-14 09:14:02 the next month so that shouldn't be an issue)

Does that sound right?

Reply:
> ignoring the doc block

Ah my bad, i tried to make sure i updated every doc block but must have missed it. I’ll fix that up so there’s no confusion

> If I remove this check

Yep. As i mentioned above, i removed this line as well and all tests passed, but then i tried to write a test to figure out why we actually need this line - thats where Test 15 came from.

In my thinking, Test #15 should be one day because if you look at the time it was first activated and when it was last deactivated, the customer only activated and deactivated across a 5 hour window, but because that 5hrs goes across 2 days in UTC (8pm to 1am) the customer gets charged for 2 days. I dont know.. maybe my thinking is off. Using a store for 5hrs and being charged two days doesn’t sound right. I do understand the 5hr window is going across two days but our other tests count two days as at least crossing over the 24hr window

It’s such an edge case but I’m just thinking if a customer somewhere in the world is activating and deactivating their store for a couple of hours every couple of days and they’ll be charged twice as much as they’re expecting

> Does that sound right?

Yeah i dont know. At the moment my preference is definitely to remove this line and then workout a fix for test 15 depending on whether it’s supposed to be one day or two days.
does that sound okay for you at this point?

Reply:
> I’ll fix that up so there’s no confusion

No dramas - I think a few might be out.

> the customer only activated and deactivated across a 5 hour window

IMO I think its 2 days. Billing isn't based on hours active in this plugin (yet). It's based on days active which have start and stop points relative to the timezone of the site.

https://github.com/prospress/woocommerce-subscriptions-resource#how-is-the-number-of-days-active-determined

> If a resource is active for more than 1 second on a given day, it will be considered to have been “active” on that day, and a prorated charge for that day will be included on the renewal.

Reply:
> our other tests count two days as at least crossing over the 24hr window

Can you point me to these ones ? You can probably spot them quicker if you been looking over them already

Reply:
Ahh ok i see where/how we're coming at this from different angles looking at the code more 🤔

Reply:
Yeah, I just pushed up another example which hopefully makes it more obvious (see 85d6654). It's the exact same total time as test 15 (I grabbed the first activation time and the last deactivated time), but I removed the multiple activations so it's just one cycle. The result is 1 day.

So essentially `Test #16` is active for longer (not deactivated across the days) and still results in 1 day (with or without removing the `--`)

> Can you point me to these ones ? You can probably spot them quicker if you been looking over them already

It's not as clear as "our other tests count two days as at least crossing over the 24hr window" but look at the doc block and code in `Test #4`. We use 1 second more then the 4 days and that checks that it's 5 days (i took that and used it for the 2 days example i.e. just over the 24hrs is counted as 2 days)